lib/sda/protocol.go

```diff
@@ -0,0 +1,179 @@
+	// Start is called when a leader has created its tree configuration and
+	// wants to start a protocol, it calls host.StartProtocol(protocolID), that
+	// in turns instantiate a new protocol (with a fresh token) , and then call
+	// Start on it.
+	Start() error
+	// Dispatch is called whenever packets are ready and should be treated
+	Dispatch([]*SDAData) error
+}
+
+// NewProtocol is the function-signature needed to instantiate a new protocol
+type NewProtocol func(*Host, *TreeNode, *Token) ProtocolInstance
+
+// ProtocolMapper handles the mapping between tokens and protocol instances. It
+// also provides helpers for protocol instances such as sending a message to
+// someone only requires to give the token and the message and protocolmapper
+// will handle the rest.
+// NOTE: This protocolMapper handle only a few things now but I suggest we leave
+// it there when our Host struct will grow. As it is starting to be already big,
+// we may, in the future, move many protocol instance /  tree / entitylist
+// handling methods in here, so host won't be too much .."overloaded" like our
+// old sign.Node. Host could relay everything that is realted to that in this
+// struct and handles the reste such as the connection, the callbacks, the
+// errors handling etc.
+type protocolMapper struct {
+	// mapping instances with their tokens
+	// maps token-uid|treenode-uid to ProtocolInstances
+	instances map[uuid.UUID]ProtocolInstance
+	// aggregate messages in order to dispatch them at once in the protocol
+	// instance
+	msgQueue map[uuid.UUID][]*SDAData
+	// Host reference
+	Host *Host
+}
+
+func newProtocolMapper(h *Host) *protocolMapper {
+	return &protocolMapper{
+		instances: make(map[uuid.UUID]ProtocolInstance),
+		msgQueue:  make(map[uuid.UUID][]*SDAData),
+		Host:      h,
+	}
+}
+
+// DispatchToInstance will dispatch this SDAData to the right instance
+// it returns true if it has successfully dispatched the msg or false
+// otherwise. It can return false because it want to aggregate some messages
+// until every children of this host has sent their messages.
+func (pm *protocolMapper) DispatchToInstance(sdaMsg *SDAData) (bool, error) {
+	var pi ProtocolInstance
+	if pi = pm.Instance(sdaMsg.To); pi == nil {
+		return false, errors.New("No instance for this token")
+	}
+	//  Get the node corresponding to this host in the Tree
+	node, err := pm.Host.TreeNodeFromToken(sdaMsg.To)
+	if err != nil {
+		return false, fmt.Errorf("Could not find TreeNode for this host in aggregate: %s", err)
+	}
+	// if message comes from parent, dispatch directly
+	if !node.IsRoot() && sdaMsg.Entity.Equal(node.Parent.Entity) {
+		return true, pi.Dispatch([]*SDAData{sdaMsg})
+	}
+
+	// if messages come from children we must aggregate them
+	var msgs []*SDAData
+	var ok bool
+	// if we still need to wait additionals message, we return
+	if msgs, ok = pm.aggregate(node, sdaMsg); !ok {
+		return false, errors.New("Still aggregating for this SDAData")
```

Comment:
Maybe this shouldn't return an `error`. The boolean flag already indicates that we aren't done aggregating messages. See https://github.com/dedis/cothority/issues/172